### ISSUE TITLE
Test Failure: QueuedThreadPoolTest.testThreadPool()

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#2619]
```
expected:<3> but was:<2>

java.lang.AssertionError: expected:<3> but was:<2>
	at org.eclipse.jetty.util.thread.QueuedThreadPoolTest.waitForThreads(QueuedThreadPoolTest.java:264)
	at org.eclipse.jetty.util.thread.QueuedThreadPoolTest.testThreadPool(QueuedThreadPoolTest.java:109)

[AdvancedRunner] Running org.eclipse.jetty.util.thread.QueuedThreadPoolTest.testThreadPool()
2018-06-01 16:00:13.729:INFO:oejut.ThreadPoolBudget:main: ReservedThreadExecutor@4e79c25{s=0/1,p=0} requires 1 threads from QueuedThreadPool[qtp1845002237]@6df87ffd{STARTING,2<=0<=4,i=0,q=0}[ReservedThreadExecutor@4e79c25{s=0/1,p=0}]
2018-06-01 16:00:13.729:WARN:oejut.ThreadPoolBudget:main: Low configured threads: (max=4 - required=1)=3 < warnAt=8 for QueuedThreadPool[qtp1845002237]@6df87ffd{STARTING,2<=0<=4,i=0,q=0}[ReservedThreadExecutor@4e79c25{s=0/1,p=0}]
```

**Comment by sbordet** [comment]
https://jenkins.webtide.net/job/jetty.project/job/jetty-9.4.x/378/testReport/junit/org.eclipse.jetty.util.thread/QueuedThreadPoolTest/linux_jdk10___Test___jdk10___testThreadPool/

**Comment by sbordet** [comment]
The reason for the failure is:
When the pool runs out of idle threads, it starts a new thread.
When the new thread starts, it first runs the idle code; if the pool is idle, then it runs the shrink code; there, it can find that the conditions to shrink are met, and immediately remove itself.
Therefore there is a race in `waitForThreads()`: waiting for the number of threads was successful, but the newly started thread was immediately removed due to shrinking, causing the assert after the wait to fail.

**Comment by sbordet** [comment]
It is not good that a newly started thread may meet the conditions to shrink and immediately be removed from the pool - this is not a good behavior of `QueuedThreadPool` hence a legit bug revealed by the test failure.

**Comment by joakime** [comment]
Backport to `jetty-9.3.x` ?

**Comment by joakime** [comment]
@sbordet this isn't a "Test" label, as the fix wasn't exclusive to Test cases only.
The fix made non-test changes in commit 77df9e1d87c10ebee18ac23af15ec4d2b379edca
